src/lib/client/jubSignalClient.ts

```diff
@@ -0,0 +1,296 @@
+  recipientPublicKey: string;
+  existingUsers: Record<string, User>;
+  existingLocationSignatures: Record<string, LocationSignature>;
+  existingActivities: Activity[];
+}): Promise<{
+  newUsers: Record<string, User>;
+  newLocationSignatures: Record<string, LocationSignature>;
+  newActivities: Activity[];
+}> => {
+  const {
+    messages,
+    recipientPrivateKey,
+    recipientPublicKey,
+    existingUsers: users,
+    existingLocationSignatures: locationSignatures,
+    existingActivities: activities,
+  } = args;
+
+  activities.reverse(); // We will reverse the activities array at the end - this is for faster array operations
+
+  for (const message of messages) {
+    let decryptedMessage: PlaintextMessage;
+    try {
+      const encryptedMessage = await encryptedMessageSchema.validate(message);
+      decryptedMessage = await decryptMessage(
+        encryptedMessage,
+        recipientPrivateKey
+      );
+    } catch (error) {
+      console.error("Invalid message received from server: ", message);
+      continue;
+    }
+
+    const { metadata, type, data } = decryptedMessage;
+
+    switch (type) {
+      case JUB_SIGNAL_MESSAGE_TYPE.OUTBOUND_TAP:
+        try {
+          if (metadata.fromPublicKey !== recipientPublicKey) {
+            throw new Error(
+              "Invalid message: outbound tap messages must be sent from self"
+            );
+          }
+
+          const { name, pk, x, tg, note } =
+            await outboundTapMessageSchema.validate(data);
+          const userId = await hashPublicKeyToUUID(pk);
+          const user = users[userId];
+          if (user) {
+            user.name = name;
+            user.encPk = pk;
+            user.x = user.x && x === undefined ? user.x : x;
+            user.tg = user.tg && tg === undefined ? user.tg : tg;
+            user.note = note;
+            user.outTs = metadata.timestamp.toISOString();
+
+            users[userId] = user;
+          } else {
+            users[userId] = {
+              name,
+              encPk: pk,
+              x,
+              tg,
+              note,
+              outTs: metadata.timestamp.toISOString(),
+            };
+          }
+
+          const activity = {
+            type: JUB_SIGNAL_MESSAGE_TYPE.OUTBOUND_TAP,
+            name,
+            id: userId,
+            ts: metadata.timestamp.toISOString(),
+          };
+          activities.push(activity);
+        } catch (error) {
+          console.error(
+            "Invalid outbound tap message received from server: ",
+            message
+          );
+        } finally {
+          break;
+        }
+      case JUB_SIGNAL_MESSAGE_TYPE.INBOUND_TAP:
+        // TODO: Can optionally validate received signature here
+        try {
+          const { x, tg, pk, msg, sig } =
+            await inboundTapMessageSchema.validate(data);
+          const userId = await hashPublicKeyToUUID(metadata.fromPublicKey);
+          const user = users[userId];
+          if (user) {
```

Comment:
Why don't we just replace user fully? Like what is the difference between the if and else here?